### ISSUE TITLE
SANDAG run time results after chunking update

### ISSUE THREAD

**Issue by esanchez01** [ActivitySim/activitysim#444]
I have been testing the recent chunking update on the 1-Zone SANDAG example. The tests were run on a server with the following specs:
- Processors: 2 Intel Xeon CPU E5-2690 v4 @ 2.60Ghz
- Cores: 28 Cores, 56 Logical Processors
- Memory: 256 GB RAM

The `chunk_training_mode`, `chunk_size`, and `num_processes` settings were set per the [documentation](https://activitysim.github.io/activitysim/core.html#chunk) -- other chunking settings were left unchanged. The sequential steps taken, and their run times are recorded in the following table:

Step | Processes | Chunk Size (RAM) | Total Run Time   (minutes) 
-- | -- | -- | --
Training_1 | 45 | 205 | 1700
Production | 45 | 205 | 573
Production | 56 | 220 | 477
Adaptive_1 | 45 | 205 | 1300
Production | 56 | 220 | 499
Adaptive_2 | 45 | 205 | 1112
Production | 56 | 230 | 529
**SERVER REBOOT**
Production | 56 | 230 | 362
Adaptive_3 | 45 | 205 | 
Production | 56 | 230 | 514
Production | 45 | 230 | 477

As can be seen, the production run times fluctuate heavily and it is hard to tell which settings are best. The best run time was achieved after a server reboot but returned to prior run times after another round of adaptive. Other processor settings were used (e.g. 28) in production runs but some memory/chunking issues were encountered (specifically at this [line](https://github.com/ActivitySim/activitysim/blob/a57efa1418320cb88d72ad5f413aaa45131183a4/activitysim/core/chunk.py#L870)).

Also, the run times we are getting here are much longer than the run times we were getting prior to the update. I compare the run times between a pre-update run using 56 processes and the best production run that achieved 362 minutes with 56 processes as well (both were run on the same machine):

model_name | Pre-Update | Post-Update | Pre - Post
-- | -- | -- | --
mandatory_tour_scheduling | 6.4 | 17.6 | -11.2
joint_tour_destination | 0.9 | 5.3 | -4.4
non_mandatory_tour_destination | 12.9 | 16 | -3.1
non_mandatory_tour_scheduling | 2.2 | 7.8 | -5.6
trip_destination | 26.8 | 65.4 | -38.6
trip_scheduling | 4.8 | 142.5 | -137.7
SUB MODEL RUN   TIME | 92.8 | 310.9 | -218.1

The main differences were kept in the table above. Trip scheduling here seems to be the main bottleneck. While comparing the log files, I noticed that 100 iterations of trip scheduling were done for each run. However, the post-update run spent a lot longer for each iteration on, what seems to be, chunking processes.

Lastly, the memory usage for the post-update production run I compared above:

<img src="https://user-images.githubusercontent.com/36172634/123495189-87ebe200-d5d7-11eb-96bb-aab655e5732e.png" width="350" />

The chunk size was set to 230 GB yet it seems to peak at ~160 GB.

**Comment by bstabler** [comment]
Hi @esanchez01.  Can you point me to your specific settings?  Maybe you can post these in a fork/branch and refer to it here?  I was getting similar/consistent runtimes in my testing so I'm not sure what's going on.  Do you know if anyone else was also using the machine?  Thanks.

**Comment by esanchez01** [comment]
@bstabler, you can find the settings/config files that were used [here](https://github.com/esanchez01/activitysim/tree/sandag-chunk-test/activitysim/examples/example_sandag/chunk_tests/sandag_1_zone_chunk). You can also find the generated chunk_cache.csv files and a subset of each run's outputs in the `test_outputs` folder. Let me know if you have any questions, thanks.

I am not aware of anyone else using the machine during the runs. After the server reboot, I did confirm that I was the only user.

**Comment by stefancoe** [comment]
@bstabler @esanchez01 we are also seeing longer run times in version 1.0.1 compared to version 0.9.9.1. I created the chunk cache following the recommended training settings and then ran in production, yielding a total runtime of 282 minutes, compared to 186 minutes in the older version.

**Comment by bstabler** [comment]
@esanchez01 - can you include the *.log files?  I wonder if the setup is thrashing?  Can you try 26 processors for your 28 core machine and set MKL threads to 1 if not previously done?  Thanks.

**Comment by esanchez01** [comment]
@bstabler, the log files are now included. For the sake of space, I include only a subset of them for each of the runs. You can find a larger subset for the last production run ([here](https://github.com/esanchez01/activitysim/tree/sandag-chunk-test/activitysim/examples/example_sandag/chunk_tests/sandag_1_zone_chunk/test_outputs/output_history/output_45proc_230ram_post_adaptive_3_server_reboot)).

I will test with 26 processors -- the MKL setting has been set for all runs.

**Comment by stefancoe** [comment]
@bstabler & @esanchez01 It looks like trip scheduling is the major culprit for us- ~60 minutes with new code compared to ~4 minutes with the old. Today was pretty busy so I did not have time to summarize run time for each sub model, but should be able to tomorrow.

**Comment by stefancoe** [comment]
For another test, I ran the old and new code using 20,000 households and turned multiprocessing off. The new code resulted in longer run times for every step (many being significantly longer). The timing logs for each run are attached. I had to stop the new run short at trip scheduling because it was taking so long, and we needed the machine for something else. It seems like there is something causing a systemic slow down for all models and trip scheduling is being hit hard because it iterates (100 times in our case) to deal with failed trips. It now takes 30-40 seconds using mp and over a minute without to run each iteration for this model, but just a few seconds in the old code.   

[timing_log.xlsx](https://github.com/ActivitySim/activitysim/files/6755391/timing_log.xlsx)

**Comment by esanchez01** [comment]
@bstabler, I had a chance to test with 26 processors. Setting the chunk size to 220-230 GB RAM, like my previous tests with 45/56 processors, lead to memory issues, as mentioned in my original post. Only after setting the chunk size to 190 GB RAM did a run complete -- anything greater failed.

The total run time for a completed run with the 190 GB RAM setting was ~360 minutes. This is similar to the best run I mentioned above but still longer than the run times achieved prior to the update. As @stefancoe mentioned, trip scheduling seems to experience the biggest slowdown.

I added the log files for both this completed run and a run that failed with a 205 GB RAM setting to my fork.

**Comment by stefancoe** [comment]
@bstabler & @esanchez01 I tested the same configuration as above (20k households, 1 processor) and applied python’s cProfile to the mandatory tour scheduling model for both new and old code. Looking at the attached results, two functions (psutil_windows.proc_memory_uss &  psutil\_pswindows.py:839(memory_full_info)) really stand out in the new code and they seem to be related to the dynamic chunking work. The rows are sorted by 'tottime', so these two are at the top in the 'new' worksheet.  I profiled the auto ownership as well and the same functions had the first and third longest total time.  I'd be happy to run it on other models.  

[mandatory_tour_scheduling_profile.xlsx](https://github.com/ActivitySim/activitysim/files/6778299/mandatory_tour_scheduling_profile.xlsx)

**Comment by stefancoe** [comment]
@bstabler Trying to figure out where psutil_windows is being called by ActivitySim. It looks like this occurs in [mem.py](https://github.com/ActivitySim/activitysim/blob/master/activitysim/core/mem.py) in several locations including [trace_memory_info](https://github.com/ActivitySim/activitysim/blob/master/activitysim/core/mem.py#L173). This module is called by [log_rss](https://github.com/ActivitySim/activitysim/blob/master/activitysim/core/chunk.py#L564) and [log_df](https://github.com/ActivitySim/activitysim/blob/master/activitysim/core/chunk.py#L583) in [chunk.py](https://github.com/ActivitySim/activitysim/blob/master/activitysim/core/chunk.py). Looking at the profile log from above, trace_memory_info is called 559 times. 

However, for that test, I used 0 chunking and did not specify a chunk mode. This seemed like the best way to compare to the old code using a small population (and not having created a training cache for such a small pop), but It turns out the code defaults to 'adpative' chunking mode when not set in the config file. @jpn--  is addressing that [here](https://github.com/ActivitySim/activitysim/pull/447#issuecomment-873097263). This would help explain why that test took so long despite running on only 20k hhs. 

So I ran a profile on school location using 'production' mode (using a training set from a full mp run) and the results are attached. Trace_memory_info is called  36 times  and psutil_windows.proc_memory_uss, which takes up the most time, is called 60 times. It seems like running in production mode should save some time, so I am going to let it run all the way through, but school location ran in 170 seconds compared to 61 with the old code, so it is still slower. 

[school_location_profile.xlsx](https://github.com/ActivitySim/activitysim/files/6780261/school_location_profile.xlsx)


EDIT:
psutil_windows.proc_memory_uss is called by this[ line](https://github.com/ActivitySim/activitysim/blob/master/activitysim/core/mem.py#L176) in mem.py. The profile log shows that this function (memory_full_info()) is called the same number of times as psutil_windows.proc_memory_uss and uses slightly more time per call.

**Comment by bstabler** [comment]
Thanks @esanchez01 for the analysis.  We agreed on the 7/8 call to further investigate.  Maybe there's too much memory use introspection still going on during a production run?  I'm going to look into the spots @stefancoe identified.

**Comment by bstabler** [comment]
I made a couple updates in this [commit](https://github.com/ActivitySim/activitysim/commit/13591836cc00507935b672d7218cebb3b7a3c321) for two issues:
  - 1 - log_rss is slow because it calls [psutil.Process.memory_full_info](https://psutil.readthedocs.io/en/latest/#psutil.Process.memory_full_info).  The linked docs say that memory_full_info is the only way to get uss and that memory_full_info is considerably slower than memory_info.  This issue can however be resolved by turning off the mem.trace_memory_info calls in production since they are not needed.  I tested with my full scale example sandag 2 zone setup from a few weeks ago and it cut the runtime from 472 minutes to 253 minutes.  
  - 2 - trip_scheduler is slow (which calls log_df (but which exits quickly if in production mode), chunk_log (which calls ChunkSizer, which calls mem.get_rss, which calls psutil.Process.memory_full_info, and adaptive_chunked_choosers_by_chunk_id (which also calls ChunkSizer).  Because ChunkSizer needs mem.get_rss, and memory_full_info is more expensive than memory_info, I tried running with ``chunk_method: hybrid_rss`` instead of ``chunk_method: hybrid_uss``.  This additionally reduced runtime to 228 minutes.  Interestingly, it reduced the trip_scheduling runtime from 79 minutes to 1.7 minutes but increased tour mode choice from 12 minutes to 58 minutes and trip mode choice from 10 minutes to 42 minutes.  

Here's the [timing_log.xlsx](https://github.com/ActivitySim/activitysim/files/6802660/timing_log.xlsx) from my sandag 2 zone testing.

**Comment by bstabler** [comment]
@esanchez01, @stefancoe - can you test the update and report your runtimes?  thanks.

**Comment by esanchez01** [comment]
@bstabler, I reran the SANDAG 1 zone example with the above changes (updates to chunk.py and using hybrid_rss as the chunk method) and achieved a total run time of ~117 minutes. This is comparable to the total run time we were seeing prior to the chunking update. The most notable differences after making the above changes are in trip scheduling, which was reduced from 142 minutes to 2 minutes, and trip destination, which was reduced from 65 minutes to 19 minutes.

[chunk_update_timing.xlsx](https://github.com/ActivitySim/activitysim/files/6804257/chunk_update_timing.xlsx) contains a run time summary for runs prior to the chunking update, with the chunking update, and after making the above changes (revised).

EDIT:
Also wanted to point out the run time increase for mandatory_tour_scheduling after the chunking update.

**Comment by stefancoe** [comment]
It now runs in 206 minutes, which is a big improvement over 282 minutes, but still a little slower than ~186 with the old code. I am going to experiment with training mode and see if I can get it down a bit more. I will also run the old code one more time to create the timing  log, which I neglected to save. 

EDIT:
It looks like trip scheduling is still longer than it was. I will try using chunk_method: hybrid_rss and report back. Also, the attached log now includes the pre-chunking release timing log. 

[timing_log_compare.xlsx](https://github.com/ActivitySim/activitysim/files/6817815/timing_log_compare.xlsx)

**Comment by bstabler** [comment]
We discussed this issue on the 7/22 call and I'm still investigating.  Turning off additional memory info calls for subsequent iterations of trip_scheduling (and just reusing the first iteration call) took the draft SANDAG two zone model from 80 minutes runtimes to 3 minutes.

**Comment by stefancoe** [comment]
@bstabler  I re-started a run after trip_purpose_and_destination with hybrid_rss and, as expected, trip scheduling ran much faster and completed within 3 minutes. Also as expected, it's now spending a lot of time running trip_mode_choice. It looks like the model is running individual chunks for each chooser:

23/07/2021 13:54:56 - INFO - activitysim.core.chunk - Running chunk 1 of 33455 with 1 of 33455 choosers
23/07/2021 13:54:59 - WARNING - activitysim.core.chunk - Not enough memory for minimum chunk_size without exceeding specified chunk_size. available_headroom: -2_961_425_660 min_chunk_size: 0 base_chunk_size: 4_106_672_900
23/07/2021 13:54:59 - INFO - activitysim.core.chunk - Running chunk 2 of 33455 with 1 of 33455 choosers
23/07/2021 13:55:03 - WARNING - activitysim.core.chunk - Not enough memory for minimum chunk_size without exceeding specified chunk_size. available_headroom: -2_961_425_660 min_chunk_size: 0 base_chunk_size: 4_106_672_900

**Comment by bstabler** [comment]
With the commit above, the in-development sandag two zone system is running about 15% slower with chunking than without chunking on my machine ([timing_log_updated.xlsx](https://github.com/ActivitySim/activitysim/files/6934419/timing_log_updated.xlsx)).  Since chunking is needed when there's not enough RAM to complete the simulation without chunking, I think this is good enough for now so I pulled these features into develop for SANDAG (@esanchez01) and PSRC (@stefancoe) to test.  I'm now working on #467 for inclusion as well.

**Comment by esanchez01** [comment]
@bstabler, I tested the SANDAG 2 zone system using develop. A few notes:

- The run was failing this [assertion](https://github.com/ActivitySim/activitysim/blob/c71e1ca56d0c6a9729fe6fd1a3922abc6a219d5d/activitysim/core/chunk.py#L738) due to the chunk tag for trip scheduling changing from `trip_scheduling_1` or `trip_scheduling_n` (previously set [here](https://github.com/ActivitySim/activitysim/commit/66d904f7b4b2f65afa3294a69569ecbdb2903d69#diff-9278809d36670ed13b751b9e2ef74910ab383cd1e148d5f5e998500e20a16903L540)) to strictly the trace label `trip_scheduling` (set [here](https://github.com/ActivitySim/activitysim/blob/c71e1ca56d0c6a9729fe6fd1a3922abc6a219d5d/activitysim/abm/models/trip_scheduling.py#L488)). This change prevented trip scheduling chunk information from being looked up in the chunk cache. After manually changing the tag in the chunk cache file, everything ran with no issues. I believe this issue arises only if the user uses an old `chunk_cache.csv` file with your update.
-  I compared the run time between the 2 zone run referenced [here](https://github.com/ActivitySim/activitysim/issues/431#issuecomment-885152285) and this new run (please note that the former incorporates some of the updates in develop). The **total** run time for the 2 zone system decreased from 389 minutes to 281. The **sub model** run time decreased from 275 minutes to 172. The scheduling, location, and destination sub models saw the most improvement.
- This may be a separate issue but notice that there is ~110 minutes gap between the sub model run time and total run time. As we have previously discussed, this may be due to the variability in run time across different processes. ActivitySim currently only logs the run time for the first process in `timing_log.csv` so it may not fully reflect the true run times of each sub model. Therefore, I added the average and max run time across processes for each sub model to the run time results. You'll notice that the sub model and total run times align better.

Here is the run time summary: [timing_log_develop.xlsx](https://github.com/ActivitySim/activitysim/files/6947606/timing_log_develop.xlsx)

**Comment by bstabler** [comment]
Yes, the old chunking file will not work since I changed the trip_scheduling chunk labeling.

I ran the 3 zone system example that you sent me on our 32 core 512 GB RAM machine and here's the 
[timing_log_updated_3zone_tests.xlsx](https://github.com/ActivitySim/activitysim/files/6949228/timing_log_updated_3zone_tests.xlsx) file.

The key numbers are:

  - 124 minutes submodel runtime with 28 processors, 500 GB chunksize, hybrid_uss
  - 102 minutes submodel runtime with 28 processors chunkless

Note also that the gap between submodel runtime and total runtime is much smaller for me.  The 124 minute submodel run total runtime was 137 minutes.  The 102 minute submodel run total runtime was 116 minutes.  

I think the difference in runtimes between your 256 GB RAM machine and my 512 GB RAM machine illustrate that these model setups need more like 512 GB of RAM.  This test also confirms that the 3 zone system setup you sent me will run without chunking on 512 GB RAM.

I'm also wondering if there are some data / logic issues in the 2 zone system setup inputs that are leading to longer runtimes than the 3 zone system setup.
